experimental/dds/tree2/docs/Shared Tree Philosophy.md

```diff
@@ -0,0 +1,204 @@
+This can get complicated, so to keep it manageable, this problem was subdivided in a few different ways.
+
+The data-model for Shared Tree was selected to assist with subdividing this problem.
+The selected data-model consists of many small structures where collaborative edits interact—fields—, and nodes which connect them into a hierarchy and assigns types to each of the fields.
+
+The nodes have schema which define the schema for their fields.
+There are a few [kinds](<https://en.wikipedia.org/wiki/Kind_(type_theory)>) of nodes that configure their fields differently, for example structs (with a fixed set of mixed field types) or maps (with an extensible set of matching field types).
+
+There are also a few kinds of fields which provide different collaborative data-structures (like `sequence`, or `optional`), each of which hold nodes.
+
+All of the editing is done in the fields via the `field kind`.
+This allows the editing logic, both API and merge policy, to be packaged together into minimal versionable units ( `field kinds`).
+Since the schema explicitly selects what field kinds to use, updating them is opt-in, and can be coordinated with deployment schedules to achieve the required cross version collaboration requirements.
+This also leverages all the same collaboration as the rest of schema evolution, making adopting alternative field kinds no different from updating other aspects of the application's schema.
+
+This results in a tree where the overall shape is controlled by the application via its tree schema (for the nodes) and the alternating layers of fields provide all the complex logic, neatly separating the concerns.
+This can also be thought of as a tree of entities alternating between fields and nodes.
+As these two kinds of entities alternate and thus never touch, this resembles a [Bipartite Graph](https://en.wikipedia.org/wiki/Bipartite_graph), and can be called a Bipartite Tree.
+
+# Constraints and Sub-Trees
+
+When users edit documents, those high level semantic operations need to be encoded in a way that supports merges.
+Applications need to be able to easily ensure the invariants they require will be maintained across merged, and (possible with some effort) high fidelity merge resolution should be achievable.
+
+While the built in merge resolution will often be sufficient to ensure the application invariants hold (see "Field Kinds" above), this won't always be the case.
+To cover those instances, an additional tool is required: "Constraints".
+Constraints are used to specify which concurrent edits could cause an unacceptable merge result.
+The constraints API (status: its not finished yet) allows defining a constraint that is sufficiently conservative, meaning that it will detect all concurrent edits that could possible be an issue, but might include some that would be fine.
```

Comment:
```suggestion
The constraints API (status: it's not finished yet) allows defining a constraint that is sufficiently conservative, meaning that it will detect all concurrent edits that could possibly be an issue, but might include some that would be fine.
```